internal/benchrunner/runners/stream/runner.go

```diff
@@ -0,0 +1,633 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package stream
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/elastic/elastic-package/internal/packages/installer"
+
+	"github.com/google/uuid"
+	"gopkg.in/yaml.v3"
+
+	"github.com/elastic/elastic-integration-corpus-generator-tool/pkg/genlib"
+	"github.com/elastic/elastic-integration-corpus-generator-tool/pkg/genlib/config"
+	"github.com/elastic/elastic-integration-corpus-generator-tool/pkg/genlib/fields"
+
+	"github.com/elastic/elastic-package/internal/benchrunner"
+	"github.com/elastic/elastic-package/internal/benchrunner/reporters"
+	"github.com/elastic/elastic-package/internal/elasticsearch"
+	"github.com/elastic/elastic-package/internal/logger"
+	"github.com/elastic/elastic-package/internal/multierror"
+	"github.com/elastic/elastic-package/internal/packages"
+	"github.com/elastic/elastic-package/internal/servicedeployer"
+	"github.com/elastic/elastic-package/internal/signal"
+)
+
+type runner struct {
+	options   Options
+	scenarios map[string]*scenario
+
+	ctxt               servicedeployer.ServiceContext
+	runtimeDataStreams map[string]string
+	generators         map[string]genlib.Generator
+	backFillGenerators map[string]genlib.Generator
+	errChanGenerators  chan error
+
+	wg   sync.WaitGroup
+	done chan struct{}
+
+	// Execution order of following handlers is defined in runner.TearDown() method.
+	removePackageHandler  func() error
+	wipeDataStreamHandler func() error
+}
+
+func NewStreamBenchmark(opts Options) benchrunner.Runner {
+	return &runner{options: opts}
+}
+
+func (r *runner) SetUp() error {
+	return r.setUp()
+}
+
+// Run runs the system benchmarks defined under the given folder
+func (r *runner) Run() (reporters.Reportable, error) {
+	return nil, r.run()
+}
+
+func (r *runner) TearDown() error {
+	r.wg.Wait()
+
+	if !r.options.PerformCleanup {
+		r.removePackageHandler = nil
+		r.wipeDataStreamHandler = nil
+		return nil
+	}
```

Comment:
I think we don't have optional cleanup in other benchmark or test runners, and this has been indeed the topic of some discussions, see https://github.com/elastic/elastic-package/issues/582

Do we need this option for some particular reason on these benchmarks?

Reply:
I pushed for this to be optional. The reason is that the `stream` command is used slightly different from other commands. A user can ship data to an existing cluster they have and will ship some data to do some testing, stop the shipping of data and then might continue again or ship some other data. If the cleanup happens, it means the user has not chance now to look at the data after stopping. It is what happened to me when I first tested. Quickly started the shipping and then wanted to look at the data without further data coming in. I stopped the shipping and suddenly my data was gone (including the package). It is also possible that for example for system, the user has some metrics already shipping but wants to add more dimension / data with the stream command.